[3단계 - 리팩터링] 도기(이동엽) 미션 제출합니다.

dye0p opened the pull request:
안녕하세요 우가! 리뷰 요청이 많이 늦어졌네요.. 3단계 구현 후 리뷰 요청합니다!
이번 단계는 객체 책임 분리에 고민을 많이 했어요. 책임 분리가 잘 되었는지 중점적으로 봐주시면 좋을 것 같습니다!
잘 부탁 드립니다~

dye0p committed:
feat: [3단계] HTTP 요청 처리 책임 분리 및 컨트롤러 기반 라우팅 구현

- FrontController 및 RequestHandler를 통해 요청 처리 구조 개선
- URL 경로 기반 컨트롤러 매핑 추가 (Controllers 및 각 컨트롤러 구현)
- LoginController 및 RegisterController 구현하여 로그인, 회원가입 처리
- Session 및 SessionManager 클래스 구현으로 세션 관리 추가
- HttpRequest, HttpResponse 클래스로 HTTP 요청 및 응답 로직 분리
- ResponseLine 클래스로 HTTP 응답 상태 코드 관리
- 정적 파일 서빙을 위한 StaticFileController 구현
- 기존 Http11Processor 로직을 컨트롤러 구조에 맞게 단순화 및 리팩터링

bowook commented on tomcat/src/main/java/org/apache/coyote/http/controller/Controller.java:
```diff
@@ -0,0 +1,9 @@
+import org.apache.coyote.http.request.HttpRequest;
+import org.apache.coyote.http.response.HttpResponse;
+
+public interface Controller {
```
Controller가 coyote에 위치하는 것이 괜찮을까요??
제가 생각하기에는, coyote는 통신의 책임이 있는 것으로 느껴져서요!

dye0p replied:
코멘트를 보고 톰캣 패키지 구조에 대해서 찾아보니 말씀하신대로 coyote 패키지는 커넥터 역할을 한다고 합니다! 

그래서 Controller가 coyote에 위치하는 것 보단 catalina에 위치하는 것이 적절해 보여요! 
그리고 Controller 구현체들은 어플리케이션 패키지(techcourse)에 들어가는게 적절할 것 같습니다!
코멘트 감사합니다 :)

[refactor: 패키지 구조 변경 및 컨트롤러 참조 경로 수정](https://github.com/woowacourse/java-http/pull/1036/commits/b06d379d9498848e9148c794a18e1382894d2591)

bowook commented on tomcat/src/main/java/org/apache/coyote/http/controller/FrontController.java:
```diff
@@ -0,0 +1,26 @@
+import org.apache.coyote.http.request.HttpRequest;
+import org.apache.coyote.http.response.HttpResponse;
+
+public abstract class FrontController implements Controller {
```
FrontController는 모든 요청을 받아서 라우팅하는 역할을 생각하고 봤는데, 실제로는 HTTP Method 처리를 서브 클래스로 위임하는 역할을 하고 있는 것 같아요.
혹시, FrontController로 지은 이유가 있으신가요!?

dye0p replied:
처음에는 Controller를 추상화하여 요청을 받아 httpMethod에 맞는 doGet, doPost를 실행시키는 구조라서 FrontController라는 이름이 자연스럽다고 생각했습니다.

하지만 코멘트를 보고 다시 생각해보니, 이 클래스는 모든 요청을 단일 지점에서 처리하는 FrontController 패턴이 아니라, 개별 컨트롤러들이 상속해서 HTTP 메서드를 나누는 추상 클래스에 더 가깝다는 걸 알게되었습니다.

오히려, 실제 FrontController 역할은 오히려 RequestHandler가 수행하고 있는 것 같네요!
그래서 이 클래스는 AbstractController라는 이름이 더 적절하다고 판단했습니다.

코멘트 덕분에 잘못 이해하고 있던 개념을 정리할 수 있었습니다. 감사합니다 :)

[refactor: FrontController 클래스명 AbstractController로 변경](https://github.com/woowacourse/java-http/pull/1036/commits/241a53b6f6e90d4e89b37b25d246fd798762603c)

bowook commented on tomcat/src/main/java/org/apache/coyote/http/controller/LoginController.java:
```diff
@@ -0,0 +1,96 @@
+    private static final String TEXT_HTML_CHARSET_UTF_8 = "text/html;charset=utf-8";
+
+    @Override
+    protected HttpResponse doGet(final HttpRequest request) throws Exception {
```
비즈니스 로직을 포함하고 있는데, was내부에 있는 것이 괜찮을까요??
이렇게 되면, 특정 애플리케이션 로직이 was와 결합될 것 같아요.

dye0p replied:
말씀해주신 것처럼 WAS와 애플리케이션 로직은 계층적으로 분리되어야 한다고 생각합니다.
그래서 비즈니스 로직을 담고 있던 객체들을 애플리케이션 패키지로 이동시켜, WAS와는 별도의 계층으로 나누었습니다!

다만 현재 컨트롤러가 여전히 HttpRequest를 매개변수로 받고 HttpResponse를 반환하고 있어서, WAS와의 결합은 남아 있습니다. 이 부분은 스프링에서 `@RequestBody,` `@ResponseBody`를 통해 DTO를 사용하는 것처럼, 중간에 `Request/Response DTO` 계층을 두어 결합을 끊는 방식으로 개선할 수 있을 것 같습니다!

[refactor: 패키지 구조 변경 및 컨트롤러 참조 경로 수정](https://github.com/woowacourse/java-http/pull/1036/commits/b06d379d9498848e9148c794a18e1382894d2591)

bowook commented on tomcat/src/main/java/org/apache/coyote/http/controller/LoginController.java:
```diff
@@ -0,0 +1,96 @@
+        String body = request.getBody();
+        final String[] queryStringParts = body.split("&");
+        final String account = queryStringParts[0].split("=")[1];
+        final String password = queryStringParts[1].split("=")[1];
```
Controller가 요청을 파싱하고 있는 것 같아요. 해당 부분은 HttpRequest의 책임으로 줄 수 있을 것 같아요!
coyote와 catalina에 대해서도 각자의 역할이 무엇인지 학습해보시면 도움이 될 것 같습니다!

dye0p replied:
그렇네요! 요청에 대한 파싱은 Controller 보단 앞단에서 해주는것이 맞는 것 같아요. 
파싱로직을 Request에서 하여 반환하도록 수정하였습니다!

[refactor: HttpRequest body 데이터 구조 Map으로 변경](https://github.com/woowacourse/java-http/pull/1036/commits/cc88b26058609485910b7bc8519e3fa496a7d386)

bowook commented on tomcat/src/main/java/org/apache/coyote/http/controller/LoginController.java:
```diff
@@ -0,0 +1,96 @@
+
+public class LoginController extends FrontController {
+
+    private static final Logger log = LoggerFactory.getLogger(Http11Processor.class);
```
로그인을 수행할 때 로깅이 되고 있지 않아용!

dye0p replied:
말씀해주신 “로그인 시 로깅”이 user 객체 출력 로그를 의미하신 걸까요?
현재는 아래처럼 로그인 성공 시 user 정보가 콘솔에 출력되도록 되어 있습니다.
<img width="1143" height="32" alt="image" src="https://github.com/user-attachments/assets/10949e7f-0e9e-4662-9350-d6a30dd47453" />

혹시 코멘트의 의도가 사용자 정보 출력이 아니라, 로그인 시도/성공/실패 이벤트 자체에 대한 로깅일까요? 제가 리뷰 의도를 제대로 이해했는지 확인 부탁드려요!

[refactor: LoginController 로깅 클래스 변경](https://github.com/woowacourse/java-http/pull/1036/commits/696607bfbba792728f450c5bf0cb475bb9e3a76d)

bowook replied:
사용자 정보 출력 로깅을 언급한 것이었는데, 잘 출력이 되는 것 같아요~

bowook commented on tomcat/src/main/java/org/apache/coyote/http/controller/LoginController.java:
```diff
@@ -0,0 +1,96 @@
+    protected HttpResponse doGet(final HttpRequest request) throws Exception {
+        final Cookie cookie = request.getCookie();
+        final Session session = SessionManager.find(cookie.getValue());
+        if (session != null) {
```
session을 만들 때 설정한 user에 대한 attribute가 null이 아닌지는 검사를 하지 않아도 괜찮을까요?

dye0p replied:
현재 비즈니스 로직상 isLoginFalied 검사를 통과한 객체를 session에 저장하고 있어서 null이 아님을 보장하고 있다고 생각했습니다. 그래서 굳이 null 검사가 필요할까? 했지만, 생각해보니 비즈니스 로직이 수정되어 null이 충분히 들어갈 수 있다고 판단했습니다. 

그래서 안정성을 위해서 말씀해주신 것 처럼 attribute에 대한 null 검사를 해주는 것이 좋을 것 같아요!

[fix: 로그인 세션 검증 로직 수정 및 Session 기능 추가](https://github.com/woowacourse/java-http/pull/1036/commits/ac7684ac8879b9d7d4f399f526ba26007b2fdc3f)

bowook commented on tomcat/src/main/java/org/apache/coyote/http/controller/LoginController.java:
```diff
@@ -0,0 +1,96 @@
+        if (isLoginFailed(user, password)) {
+            try {
+                final URL resource = getClass().getClassLoader().getResource("static" + "/401" + ".html");
+                validateNullResource(resource);
```
전체적으로 컨트롤러에서 리소스 존재 여부를 직접 체크하고 있는 부분이 있는데, 
이걸 컨트롤러보다는 파일 경로를 파싱하고 리소스를 확인해주는 별도 객체에 맡기면 어떨까 싶어요. 
컨트롤러는 HTTP 요청/응답 흐름만 담당하도록 하면 책임이 더 명확해질 것 같습니다.

dye0p replied:
좋은것 같아요! 별도의 객체로 분리하였습니다~

[refactor: ResourceResolver 도입 및 리소스 처리 로직 개선](https://github.com/woowacourse/java-http/pull/1036/commits/122809053ba8767e7632cf51ed0477d3c03b42de)

bowook commented on tomcat/src/main/java/org/apache/coyote/http/controller/RegisterController.java:
```diff
@@ -0,0 +1,48 @@
+
+    private User createUser(String account, String password, String email) {
+        Long id = 1L;
+        return new User(++id, account, password, email);
```
User의 id를 증가시키는 부분이 지금 Controller에 있는데, 
실제로 User를 관리하는 곳은 InMemoryUserRepository라서 RegisterController의 책임이 조금 더 있는 것 같아요!
그리고, 아래 코드래빗의 리뷰도 참고해보시면 좋을 것 같습니다!
https://github.com/woowacourse/java-http/pull/1036/files#r2343135151

dye0p replied:
로직 이동을 고민했지만, 조금 생각해보니 아직은 user의 id를 증가시킬 필요가 굳이 없는 것 같아서 id증가 시키는 부분을 제거하고 기본 생성자로 user를 생성하도록 변경하였습니다!

[refactor: RegisterController에서 불필요한 createUser 메서드 제거](https://github.com/woowacourse/java-http/pull/1036/commits/2cbde25f4f3af9f50e4ec257e4acdcfcce2cf7f9)

bowook commented on tomcat/src/main/java/org/apache/coyote/http/controller/StaticFileController.java:
```diff
@@ -0,0 +1,33 @@
+
+    private void validateNullResource(final URL resource) {
+        if (resource == null) {
+            throw new IllegalArgumentException("존재하지 않는 resource 입니다.");
```
존재하지 않는 리소스에 대해서 검증하는 로직이 각 구체 컨트롤러마다 존재하는 것 같아요.
경로를 파싱하는 객체를 뽑아내서 그 객체에게 책임을 인가 할 수 있을 것 같아요!

dye0p replied:
경로를 파싱하는 객체를 분리하면서 함께 추가하였습니다!

[refactor: ResourceResolver 도입 및 리소스 처리 로직 개선](https://github.com/woowacourse/java-http/pull/1036/commits/122809053ba8767e7632cf51ed0477d3c03b42de)

bowook replied:
리졸버를 도입하셨군요!! 👍

bowook commented on tomcat/src/main/java/org/apache/coyote/http/request/RequestParser.java:
```diff
@@ -0,0 +1,56 @@
+        return getRequestBody(contentLength, bufferedReader);
+    }
+
+
```
개행이 두 줄 되어있어요!
전체적으로 다른 클래스들도 확인하시면 좋을 것 같아요!

dye0p replied:
확인하였습니다!

bowook commented on tomcat/src/main/java/org/apache/coyote/http11/Http11Processor.java:
```diff
@@ -48,213 +39,19 @@ public void process(final Socket connection) {
-    }
-
-    private void writeAndFlush(final OutputStream outputStream, final String response) throws IOException {
+    public void writeAndFlush(final OutputStream outputStream, final String response) throws IOException {
```
기존에 private에서 public으로 바꾸신 이유가 있으신가요?

dye0p replied:
리팩터링을 하면서 다른 클래스에 위치했다가 다시 돌아오면서 public으로 바꿨던걸 못봤네요! private가 맞습니다~

[refactor: 접근제어자 변경(public -> private)](https://github.com/woowacourse/java-http/pull/1036/commits/54b0108f3e99b030e245a4354f4586da1807f997)

bowook commented on tomcat/src/main/java/org/apache/coyote/http/response/HttpResponse.java:
```diff
@@ -0,0 +1,64 @@
+        return responseBuilder.toString();
+    }
+
+    public String getBody() {
```
사용하지 않는 메서드인 것 같습니다!

dye0p replied:
제거하였습니다!

[refactor: 사용하지 않는 메서드 제거](https://github.com/woowacourse/java-http/pull/1036/commits/68e7a70823a81c1f02ada45bfd320449916bc65a)

bowook commented on tomcat/src/main/java/org/apache/coyote/http/request/HttpRequest.java:
```diff
@@ -0,0 +1,47 @@
+    private final RequestLine requestLine;
+    private final Map<String, String> headers;
+    private final String body;
+    private final Cookie cookie;
```
HttpRequest에서 쿠키를 별도로 관리하고 있는데, 원래 쿠키는 요청 헤더에 포함되는 것이라
이렇게 따로 관리하면 관리 포인트가 늘어나는 느낌이 있어요. 
특히 요청 헤더에 쿠키가 없으면 emptyCookie를 만들어 넣어주고 있는데, 
혹시 이 부분도 꼭 별도로 관리해야 하는지 고민해볼 수 있을 것 같아요.

dye0p replied:
헤더에 쿠키가 들어가는 것을 인지하지 못했네요.. 그럼  HttpRequest에 cookie 필드가 필요없어 보입니다!
> 특히 요청 헤더에 쿠키가 없으면 emptyCookie를 만들어 넣어주고 있는데,
혹시 이 부분도 꼭 별도로 관리해야 하는지 고민해볼 수 있을 것 같아요.

이 부분은 현재 Cookie 객체를 반환하는 로직상 쿠키가 존재하지 않는다면 빈 쿠키를 반환하거나, null을 반환해야 하는데,
이 부분은 Optional을 통해서 해결하였습니다!

[refactor: Request 쿠키 파싱 및 세션 검증 로직 개선](https://github.com/woowacourse/java-http/pull/1036/commits/172116c646eed737783cf9928acaf831cf491e2d)

bowook reviewed (changes requested):
도기, 안녕하세요~
깔끔하게 잘 구현해주신 것 같아요!
궁금한 점 몇 가지와 제안해보고 싶은 점 코멘트 남겨봤습니다. 확인 부탁드릴게요!

dye0p committed:
refactor: 패키지 구조 변경 및 컨트롤러 참조 경로 수정

- 모든 컨트롤러 패키지를 com.techcourse.controller로 이동
- FrontController 및 Controller 인터페이스 경로를 org.apache.catalina.controller로 변경
- RequestHandler 클래스에서의 컨트롤러 참조 경로 수정
- 사용되지 않는 import 제거 및 정리

dye0p committed:
refactor: FrontController 클래스명 AbstractController로 변경

- 모든 컨트롤러에서 FrontController를 AbstractController로 변경
- 클래스 및 import 경로 수정
- 추상 클래스 명칭을 역할에 맞게 더 명확히 변경

dye0p committed:
fix: 로그인 세션 검증 로직 수정 및 Session 기능 추가

- validateSession 메서드 추가로 세션 검증 로직 캡슐화
- Session 클래스에 getAttribute 메서드 추가로 속성 조회 가능하도록 구현

dye0p committed:
refactor: HttpRequest body 데이터 구조 Map으로 변경

- HttpRequest의 body 데이터 타입을 String에서 Map으로 수정
- RequestParser에서 body 파싱 로직 변경 (key-value 형태로 변환)
- HttpRequest#getBody가 불변 Map 반환하도록 수정
- LoginController 및 RegisterController의 요청 본문 처리 로직 간소화
- getRequestBody 및 parseBody 메서드 네이밍 및 역할 개선

dye0p committed:
refactor: ResourceResolver 도입 및 리소스 처리 로직 개선

- ResourceResolver 클래스를 추가하여 중복된 리소스 처리 로직 제거
- 리소스 확장자 패턴 매칭 기능을 통해 기본 확장자(.html) 자동 추가
- validateNullResource 메서드 ResourceResolver로 이동하여 코드 재사용성 향상
- LoginController, RegisterController 등에서 리소스 로딩 로직 ResourceResolver로 통합
- 불필요한 validateNullResource 중복 메서드 제거

dye0p committed:
refactor: 사용하지 않는 메서드 제거

dye0p committed:
refactor: 접근제어자 변경(public -> private)

dye0p committed:
refactor: RegisterController에서 불필요한 createUser 메서드 제거

- User 객체 생성 로직을 생성자 호출로 단순화
- 불필요한 메서드 제거로 코드 가독성 향상

dye0p committed:
refactor: Request 쿠키 파싱 및 세션 검증 로직 개선

- Cookie 파싱 메서드 반환 타입 Optional로 변경
- HttpRequest에서 Cookie 필드 제거 및 헤더 기반 Cookie 파싱 로직 추가
- LoginController에서 쿠키 검증 및 세션 리다이렉트 로직 간소화
- Session 클래스에 isValid 메서드 추가로 세션 검증 캡슐화
- 중복된 리소스 로딩 로직을 별도 메서드로 분리

dye0p committed:
refactor: LoginController 로그인 실패 시 리다이렉트 로직 간소화

- 리소스 로드 및 파일 읽기 대신 단순 리다이렉션으로 수정
- HttpResponse.redirection 호출로 코드 가독성 및 유지보수성 향상

dye0p committed:
refactor: LoginController 로깅 클래스 변경

- Logger 참조를 Http11Processor에서 LoginController로 수정
- 로깅 대상 클래스 명확화로 코드 일관성 개선

dye0p committed:
refactor: ResourceResolver 패키지 경로 변경

- org.apache.catalina에서 org.apache.catalina.request로 이동
- 관련된 import 경로 수정 (LoginController, RegisterController 등)

dye0p reviewed (commented):
안녕하세요 우가! 리뷰 반영이 조금 걸렸네요 좋은 코멘트를 많이 남겨주셔서 많이 알게된 것 같습니다 감사합니다 :)

dye0p committed:
refactor: LoginController 리다이렉트 URL 경로 수정

- index.html 리소스 접근 경로를 "/index.html"로 변경
- 리다이렉트 경로 명확화로 코드 일관성 개선

bowook reviewed (approved):
고생하셨습니다!
빠르게 코멘트 반영해주시고 생각 남겨주셔서 같이 고민해 볼 수 있었던 것 같아요!
💯
